### PR TITLE
commit: add --max-layers argument

### PR DESCRIPTION
**Pull request by TristanCacqueray**
Fixes: #1778

### CODE REVIEW

**Comment by TristanCacqueray** [comment]
I'm not very familiar with the code base, but that change seems to work. Would that be the right strategy to implement #1778 @TomSweeneyRedHat ?

**Comment by TristanCacqueray** [comment]
This change needs more assertion to prevent an invalid max-layers argument, and the new history order may be incorrect.

File: docs/buildah-commit.md [hunk 1]
```diff
@@ -64,7 +64,11 @@ Default leaves the container and its content in place.
 
 **--squash**
 
-Squash all of the new image's layers (including those inherited from a base image) into a single new layer.
+Squash all of the new image's layers (including those inherited from a base image** into a single new layer.
```
Comment (rhatdan) [inline docs/buildah-commit.md]: Should conflict with --squash

File: image.go [hunk 2]
```diff
@@ -180,8 +181,14 @@ func (i *containerImageRef) createConfigsAndManifests() (v1.Image, v1.Manifest,
 	oimage.RootFS.DiffIDs = []digest.Digest{}
 	// Only clear the history if we're squashing, otherwise leave it be so that we can append
 	// entries to it.
-	if i.squash {
-		oimage.History = []v1.History{}
+	if i.squash || i.squashMaxLayers > 0 {
+		newHistory := []v1.History{}
+		for idx, hist := range oimage.History {
```
Comment (rhatdan) [inline image.go]: Should only run this code if !i.squash

File: image.go [hunk 3]
```diff
@@ -198,8 +205,14 @@ func (i *containerImageRef) createConfigsAndManifests() (v1.Image, v1.Manifest,
 	dimage.RootFS.DiffIDs = []digest.Digest{}
 	// Only clear the history if we're squashing, otherwise leave it be so that we can append
 	// entries to it.
-	if i.squash {
-		dimage.History = []docker.V2S2History{}
+	if i.squash || i.squashMaxLayers > 0 {
+		newHistory := []docker.V2S2History{}
+		for idx, hist := range dimage.History {
```
Comment (rhatdan) [inline image.go]: Should not run this code if i.squash.

**Comment by rhatdan** [comment]
Not hugely familiar with this code, but should --squash set the --squash-max-layers flag to1?

File: docs/buildah-commit.md (hunk 1 above)
Comment (TristanCacqueray) [reply docs/buildah-commit.md]: Commented

File: image.go (hunk 2 above)
Comment (TristanCacqueray) [reply image.go]: Fixed

File: image.go (hunk 3 above)
Comment (TristanCacqueray) [reply image.go]: Fixed

**Comment by TristanCacqueray** [comment]
Last commit renamed the argument to "--max-layers" as it feels more natural.

**Comment by TristanCacqueray** [comment]
Last commit fixed the history order. Please let me know if the implementation is correct or how can it be improved.

File: cmd/buildah/commit.go [hunk 4]
```diff
@@ -97,6 +99,12 @@ func commitCmd(c *cobra.Command, args []string, iopts commitInputOptions) error
 	if err := buildahcli.VerifyFlagsArgsOrder(args); err != nil {
 		return err
 	}
+	if iopts.maxLayers != ^uint(0) && iopts.squash && iopts.maxLayers != 1 {
+		return errors.Errorf("--squash and --max-layers are mutually exclusive")
+	}
+	if iopts.maxLayers == 0 {
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/commit.go]: pref "< 1"

File: image.go [hunk 5]
```diff
@@ -168,6 +169,11 @@ func (i *containerImageRef) extractRootfs() (io.ReadCloser, error) {
 func (i *containerImageRef) createConfigsAndManifests() (v1.Image, v1.Manifest, docker.V2Image, docker.V2S2Manifest, error) {
 	created := i.created
 
+	if i.maxLayers == 0 {
```
Comment (TomSweeneyRedHat) [inline image.go]: pref "< 1"

File: image.go [hunk 6]
```diff
@@ -229,6 +253,11 @@ func (i *containerImageRef) createConfigsAndManifests() (v1.Image, v1.Manifest,
 }
 
 func (i *containerImageRef) NewImageSource(ctx context.Context, sc *types.SystemContext) (src types.ImageSource, err error) {
+	if i.maxLayers == 0 {
```
Comment (TomSweeneyRedHat) [inline image.go]: pref "< 1"

File: image.go [hunk 7]
```diff
@@ -344,6 +400,15 @@ func (i *containerImageRef) NewImageSource(ctx context.Context, sc *types.System
 			if err != nil {
 				return nil, err
 			}
+		} else if toBeSquashed(layerID) {
+			// Extract all the extra layers into a single diff
+			firstID := layersToSquash[len(layersToSquash)-1]
+			lastID := layersToSquash[0]
+			rc, err = i.store.Diff(firstID, lastID, diffOptions)
+			logrus.Debugf("squashing old layers from %q to %q", firstID, lastID)
```
Comment (TomSweeneyRedHat) [inline image.go]: :+1:

File: cmd/buildah/commit.go (hunk 4 above)
Comment (TristanCacqueray) [reply cmd/buildah/commit.go]: Done

File: image.go (hunk 5 above)
Comment (TristanCacqueray) [reply image.go]: Done

File: image.go (hunk 6 above)
Comment (TristanCacqueray) [reply image.go]: Done

**Comment by TristanCacqueray** [comment]
Thanks for the review, I'll push a few more commits soon.

**Comment by TristanCacqueray** [comment]
Turns out `store.Diff()` doesn't return the from layer content when from != to.Parent, thus this only work when committing an image built `from scratch`. When testing with `from image`, the resulting commit is missing the original from layer.

It seems like the storage driver operate differently when from != to.Parent and I'm not sure how to produce a diff that includes the from layer.

**Comment by TristanCacqueray** [comment]
I couldn't find why `store.Diff()` doesn't work as expected. Actually `store.Mount()` work better and https://github.com/containers/buildah/pull/1790/commits/999a5cb2a76236304891cd790e424cc5fada7e5a refactor the logic a bit as mounting is easier to use: only the top layerID needs to be tracked.

With this last update, I was able to squash a 50 layers (6GB) fedora image down to 8 layers (4.5GB).

**Comment by TristanCacqueray** [comment]
@rhatdan the PR is working as expected. Though it could use a squash, I left the `store.Mount` fix in a separate commit on top as @TomSweeneyRedHat only reviewed the `store.Diff` strategy.

I think we are also waiting for @nalind feedback. And I'd be happy to follow-up suggestions.

**Comment by TomSweeneyRedHat** [comment]
@nalind, could you peruse this please and we'll merge if it LGTY?

**Comment by TristanCacqueray** [comment]
Perhaps `--max-layers` is not a good name. There actually multiple strategies that could be considered:

* [squash from 1 to n-max] + n-max..n   (the current implementation, discard the oldest layer, useful for rolling update of an image)
* [1] + [squash from 2 to n-max-1] + n-max-1..n   (to keep the initial layer) 
* 1..n-max + [squash from n-max to n]    (to keep the oldest, discard the newest, useful for rebasing updates to a multi-layer images)

**Comment by TomSweeneyRedHat** [comment]
@TristanCacqueray needs a rebase

**Comment by TristanCacqueray** [comment]
@TomSweeneyRedHat alright, rebase was trivial

**Comment by TomSweeneyRedHat** [comment]
@nalind, @rhatdan , @vrothberg PTAL
Thanks @TristanCacqueray !

**Comment by TomSweeneyRedHat** [comment]
@nalind @rhatdan @vrothberg RE-PING!

**Comment by TristanCacqueray** [comment]
Actually I think the second use-case mentioned in https://github.com/containers/buildah/pull/1790#issuecomment-534579845 would be more useful. On a big image, the current implementation is quite slow when the squash happens. I think this review shouldn't touch the initial layer, and let the user manage it manually, for example by using `--squash` periodically.

**Comment by rhatdan** [comment]
@nalind @mtrmac @vrothberg @giuseppe This is an old PR but might be worth moving forward on.

**Comment by rhatdan** [comment]
Since this PR has not been touched in a year, I am going to close.  Reopen if more work is going to be done on it.

**Pull request closed without merging**